### ISSUE TITLE
test ./imjournal-basic.sh frequenly fails

### ISSUE THREAD

**Issue by rgerhards** [rsyslog/rsyslog#2564]
currently seen very frequently on fedora buildbot slave.

**recorded failures: http://build.rsyslog.com/testbench-showtest.php?testname=imjournal-basic%.sh**

The failure looks a bit different, here is a frequent case:
```
------------------------------------------------------------
Test: ./imjournal-basic-vg.sh
------------------------------------------------------------
testbench: TZ env var not set, setting it to UTC
testconf.conf is:
     1	$IncludeConfig diag-common.conf
     2	
     3	module(load="../plugins/imjournal/.libs/imjournal")
     4	
     5	template(name="outfmt" type="string" string="%msg%\n")
     6	action(type="omfile" template="outfmt" file="rsyslog.out.log")
==4868== Memcheck, a memory error detector
==4868== Copyright (C) 2002-2017, and GNU GPL'd, by Julian Seward et al.
==4868== Using Valgrind-3.13.0 and LibVEX; rerun with -h for copyright info
==4868== Command: ../tools/rsyslogd -C -n -irsyslog.pid -M../runtime/.libs:../.libs -ftestconf.conf
==4868== 
rsyslogd started, start msg not yet seen, pid  4868
rsyslogd started, start msg not yet seen, pid  4868
rsyslogd startup msg seen, pid  4868
startup-vg still running
Written to journal: "TestBenCH-RSYSLog imjournal This is a test message - 1521131225"
imdiag[13500]: mainqueue empty
==4868== 
==4868== HEAP SUMMARY:
==4868==     in use at exit: 24,883 bytes in 56 blocks
==4868==   total heap usage: 4,536,637 allocs, 4,536,581 frees, 2,267,745,613 bytes allocated
==4868== 
==4868== 263 (88 direct, 175 indirect) bytes in 1 blocks are definitely lost in loss record 37 of 42
==4868==    at 0x4C2FA1E: calloc (vg_replace_malloc.c:711)
==4868==    by 0xBC6AC27: nspollConstruct (nspoll.c:93)
==4868==    by 0xBE75C6E: Run (tcpsrv.c:919)
==4868==    by 0x46A260: thrdStarter (threads.c:223)
==4868==    by 0x52D936C: start_thread (in /usr/lib64/libpthread-2.25.so)
==4868==    by 0x6226B4E: clone (in /usr/lib64/libc-2.25.so)
==4868== 
{
   <insert_a_suppression_name_here>
   Memcheck:Leak
   match-leak-kinds: definite
   fun:calloc
   fun:nspollConstruct
   fun:Run
   fun:thrdStarter
   fun:start_thread
   fun:clone
}
==4868== LEAK SUMMARY:
==4868==    definitely lost: 88 bytes in 1 blocks
==4868==    indirectly lost: 175 bytes in 4 blocks
==4868==      possibly lost: 0 bytes in 0 blocks
==4868==    still reachable: 24,620 bytes in 51 blocks
==4868==         suppressed: 0 bytes in 0 blocks
==4868== Reachable blocks (those to which a pointer was found) are not shown.
==4868== To see them, rerun with: --leak-check=full --show-leak-kinds=all
==4868== 
==4868== For counts of detected and suppressed errors, rerun with: -v
==4868== ERROR SUMMARY: 1 errors from 1 contexts (suppressed: 0 from 0)
rsyslogd run exited with 10
valgrind run FAILED with exceptions - terminating
FAIL imjournal-basic-vg.sh (exit status: 1)
```
In other cases we see connection errors, which might be a startup issue or extremely slow machine (very unlikely).

**Comment by rgerhards** [comment]
as this seems to be more complex to solve, we temporarily disable imjournal tests via #2567

**Comment by rgerhards** [comment]
One thing I do not understand is that, with the possibly fixed tests, this appears:
```
------------------------------------------------------------
Test: ./imjournal-basic.sh
------------------------------------------------------------
testbench: TZ env var not set, setting it to UTC
testconf.conf is:
     1	$IncludeConfig diag-common.conf
     2	
     3	module(load="../plugins/imjournal/.libs/imjournal")
     4	
     5	template(name="outfmt" type="string" string="%msg%\n")
     6	action(type="omfile" template="outfmt" file="rsyslog.out.log")
rsyslogd started, start msg not yet seen, pid  30647
rsyslogd startup msg seen, pid  30647
Written to journal: "TestBenCH-RSYSLog imjournal This is a test message - 1521477857"
imdiag[13500]: mainqueue empty
No journal files were found.
FAIL: rsyslog.out.log content (tail -n200):
 [origin software="rsyslogd" swVersion="8.34.0.master" x-pid="30647" x-info="http://www.rsyslog.com"] start
 [origin software="rsyslogd" swVersion="8.34.0.master" x-pid="30647" x-info="http://www.rsyslog.com"] exiting on signal 15.
=======
last entries from journal:
No journal files were found.
=======
NOTE: last 200 lines may be insufficient on busy systems!
=======
FAIL: imjournal test message could not be found!
Expected message content was:
TestBenCH-RSYSLog imjournal This is a test message - 1521477857
```

That's on the CentOS 7 buildbot slave. I do not understand why we do not have the data inside the journal, at least in temp files. I have reconfigured buildbot to not build imjournal (and thus do the tests) on the slaves that have this issue. Doesn't look like the right solution, though - but probably better than when I really need to fully disable the test. @Whissi @mbiebl do you have an idea what causes this and how I could work around it?

**Comment by Whissi** [comment]
> FAIL: imjournal test message could not be found!

Do we know if the test message wasn't recorded at all or if you hit the 200 lines limit?

At least the check from *journal_print* is not enough, from [sd_journal_print](https://www.freedesktop.org/software/systemd/man/sd_journal_print.html):
> If systemd-journald(8) is not running (the socket is not present), those functions do nothing, and also return 0.

Would be interesting to see how *systemd-cat* performs instead.

**Comment by bluca** [comment]
Maybe I'm missing something, but shouldn't the test check for the message in rsyslog.out.log where the im module should have forwarded it to? Given it uses journalctl, as far as I can see at the moment it's just testing that the journal_print helper program injects the message in the journal?

**Comment by rgerhards** [comment]
@bluca correct comment, unfortunately I had not time back in may to look at it. Thankfully #2851 fixes that issue. Now we need to see if problems with the journal tests persists or are solved.

**Comment by bluca** [comment]
Naive question: is "startup" synchronous or asynchronous? If the latter, could it be a race between the script sending the message and rsyslogd starting, since the module is using IgnorePreviousMessages="on" ?

**Comment by rgerhards** [comment]
> Naive question: is "startup" synchronous or asynchronous?

synchronous

I work again on this test, at least trying to make it skip instead of fail if the message does not show up in the journal. I'll also review the test method in depth just to be sure.

**Comment by jvymazal** [comment]
> I work again on this test, at least trying to make it skip instead of fail if the message does not show up in the journal.

@rgerhards, AFAIK the test **does** skip if the message is not present in the journal. The last log I tried to decipher was that rsyslog exited long before the message appeared in the journal. However I am not familiar with the testbench enough to propose a fix for that.

**Comment by rgerhards** [comment]
> The last log I tried to decipher was that rsyslog exited long before the message appeared in the journal.

aha! any evidence of this would be helpful - but this points into a direction I am also currently looking. The sync is currently done via a sleep and via the regular shutdown logic. All in all, this gives the journal roughly 1.5 sec to forward the message before rsyslog shuts down. So this actually **is** a race, and would match what you describe. I will add some logic that waits longer.

**Comment by jvymazal** [comment]
> any evidence of this would be helpful 

Ok, if you look for example at https://build.rsyslog.com/#/builders/175/builds/660/steps/6/logs/test-suite_log and go trough it you can see this lines in the journal cat:
```
Oct 26 13:32:11 centos7-5-w3-1vcpu-1gb-fra1-02 rsyslogd[17327]: imfile: on startup file '/var/lib/buildbot/slave/rsyslog_centos7-5/build/tests/rstb_848663_1ee4d6b7.input' does not exist but is configured in static file monitor - this may indicate a misconfiguration. If the file appears at a later time, it will automatically be processed. Reason: No such file or directory [v8.39.0.master]
Oct 26 13:32:11 centos7-5-w3-1vcpu-1gb-fra1-02 rsyslogd[17327]:  [origin software="rsyslogd" swVersion="8.39.0.master" x-pid="17327" x-info="http://www.rsyslog.com"] start
Oct 26 13:32:11 centos7-5-w3-1vcpu-1gb-fra1-02 rsyslogd[17327]:  [origin software="rsyslogd" swVersion="8.39.0.master" x-pid="17327" x-info="http://www.rsyslog.com"] exiting on signal 15.
Oct 26 13:36:05 centos7-5-w3-1vcpu-1gb-fra1-02 imjournal_test[22377]: TestBenCH-RSYSLog imjournal This is a test message - 1540560965 - rstb_110547_944b496a
```
I do not know how busy can journal be, however here it was several *minutes* before the message got trough.

**Comment by rgerhards** [comment]
wow - you are right! 4 minutes latency... ummm; that's a different kind of timeout compared to what I thought about- but it looks like we have a path forward. Maybe try for 5 minutes to hope the journal sees the message and exit 77 if not?

**Comment by rgerhards** [comment]
ok, I changed the test in #3245 to wait 5 minutes for the journal to forward the message. Let's see how it works out. Any more comments, including on the PR, are very welcome.

**Comment by rgerhards** [comment]
side-note: -3245 is a prototype PR. If it looks like it works out, I'll further improve on it. Think about adding plumbing that other journal tests could also utilize.

**Comment by Whissi** [comment]
I am wondering if sending `SIGUSR1` or calling `journalctl --flush` will help here...

**Comment by rgerhards** [comment]
I have merged the experimental PR - let's see how this works out the next days.

One interesting finding: there is a related valgrind-test. That failed very infrequently (if at all). So it looks like the small delay caused by valgrind was sufficient to make the test succeed. That's of course counter the idea of minute-long latency. 

Not just out of curiosity in the follow-up PR I am working on I have done the journalctl check directly after the message injection. And interestingly, journalctl sees the message immediately. @Whissi I would also think that this is an indicaton ' journalctl --flush' is not necessary. I'll keep on eye on this test. Given the rate with with it failed, it should become fairly soon clear if we went towards a solution.
